bpo-41086: Add exception for uninstantiated interpolation (configparser)

scrummyin opened the pull request:
The current feedback when users try to pass an uninstantiated
interpolation into a ConfigParser is an error message that does not help
users solve the problem. This current error of `TypeError: before_set()
missing 1 required positional argument: 'value'` does not display until
the parser is used, which usually results in the assumption that
instantiation of the parser was done correctly. The new exception of
InterpolationIsNotInstantiatedError, will be raised on the line where
the ConfigParser is instantiated. This will result in users see the line
that has the error in their backtrace for faster debugging.

There have been a number of bugs created in the issue tracker, which
could have been addressed by:
https://bugs.python.org/issue26831 and https://bugs.python.org/issue26469


<!-- issue-number: [bpo-41086](https://bugs.python.org/issue41086) -->
https://bugs.python.org/issue41086
<!-- /issue-number -->

scrummyin committed:
Add exception for uninstantiated interpolation (configparser)

The current feedback when users try to pass an uninstantiated
interpolation into a ConfigParser is an error message that does not help
users solve the problem. This current error of `TypeError: before_set()
missing 1 required positional argument: 'value'` does not display until
the parser is used, which usually results in the assumption that
instantiation of the parser was done correctly. The new exception of
InterpolationTypeError, will be raised on the line where the
ConfigParser is instantiated. This will result in users see the line
that has the error in their backtrace for faster debugging.

There have been a number of bugs created in the issue tracker, which
could have been addressed by:
https://bugs.python.org/issue26831 and https://bugs.python.org/issue26469

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

iritkatriel commented on Lib/configparser.py:
```diff
@@ -633,6 +642,8 @@ def __init__(self, defaults=None, dict_type=_default_dict,
         if self._interpolation is None:
             self._interpolation = Interpolation()
+        if isinstance(self._interpolation, type):
+            raise InterpolationIsNotInstantiatedError()
```
If the parameter is to be checked, then I think the check should be that interpolation is an instance of Interpolation (rather than that is is not an instance of type). That way you would catch other bad args.

The exception would be InterpolationTypeError or something like that.

scrummyin replied:
Thanks for the suggestions, I've accepted them locally, but they cause some need for changes in the English text/explanations associated with the code. I'll made updates to those and post back once I get better text in there.

scrummyin replied:
In response to the feedback, the exception now is InterpolationTypeError and some of the comment text and the commit message has been updated to reflect this. The patch was also rebased onto main as there was a comment on https://bugs.python.org/issue41086 about not backporting and a conflict in the `__all__` which the rebase has resolved.

scrummyin replied:
@iritkatriel , sorry to send a notification, but I wanted to check if I corrected things per your suggestion?

iritkatriel commented on Lib/configparser.py:
```diff
@@ -279,6 +279,13 @@ class InterpolationSyntaxError(InterpolationError):
+
+    Giving users faster and more clear feedback on how to fix a common error
+    """
+    def __init__(self):
```
The other exceptions in the module call Error.__init__, this one should follow this pattern.

iritkatriel commented on Lib/configparser.py:
```diff
@@ -279,6 +279,13 @@ class InterpolationSyntaxError(InterpolationError):
+class InterpolationTypeError(Error):
+    """Raised when the _interpolation is not an instance of Interpolation
+
+    Giving users faster and more clear feedback on how to fix a common error
```
The last line of this docstring isn't really providing useful information.

iritkatriel commented on Lib/test/test_configparser.py:
```diff
@@ -1048,6 +1048,12 @@ def test_set_malformatted_interpolation(self):
         self.assertEqual(cf.get("sect", "option2"), "foo%%bar")
 
 
+class ConfigParserTestCaseInterpolationIsInstantiated(unittest.TestCase):
```
I'd make this ConfigParserTestCaseInvalidInterpolationType   and add a couple of tests for other invalid types (say, an int and a string. You can do this in a loop), something like:
```

for value in [configparser.ExtendedInterpolation,  42,  "a string]:
   with self.subTest(value = value):
        with self.assertRaises(TypeError):
              configparser.ConfigParser(interpolation=value)
```

iritkatriel commented on Lib/configparser.py:
```diff
@@ -279,6 +279,13 @@ class InterpolationSyntaxError(InterpolationError):
     which substitutions are made does not conform to the required syntax.
     """
 
+class InterpolationTypeError(Error):
```
If you add a new exception type then you need to add it to the documentation as well. TBH, I don't think you need a new exception type here. You could just raise a TypeError.

iritkatriel commented on Lib/test/test_configparser.py:
```diff
@@ -1048,6 +1048,12 @@ def test_set_malformatted_interpolation(self):
 
+class ConfigParserTestCaseInterpolationIsInstantiated(unittest.TestCase):
+    def test_error_on_uninstantiated_interpolation(self):
+        with self.assertRaises(configparser.InterpolationTypeError) as context:
```
```suggestion
        with self.assertRaises(configparser.InterpolationTypeError):
```

iritkatriel commented on Lib/test/test_configparser.py:
```diff
@@ -1048,6 +1048,12 @@ def test_set_malformatted_interpolation(self):
+class ConfigParserTestCaseInterpolationIsInstantiated(unittest.TestCase):
+    def test_error_on_uninstantiated_interpolation(self):
+        with self.assertRaises(configparser.InterpolationTypeError) as context:
+            cf = configparser.ConfigParser(interpolation=configparser.ExtendedInterpolation)
```
```suggestion
            configparser.ConfigParser(interpolation=configparser.ExtendedInterpolation)
```

iritkatriel commented on Misc/NEWS.d/next/Library/2020-06-23-01-50-24.bpo-41086.YnOvpS.rst:
```diff
@@ -0,0 +1 @@
+Added validation check of instantiation on ConfigParser's interpolation parameter.
```
```suggestion
Make the :class:`configparser.ConfigParser` constructor raise :exc:`TypeError` if the ``interpolation`` parameter is not of type :class:`configparser.Interpolation`.
```

scrummyin committed:
Replace custom Error with TypeError

Per feedback from @iritkatriel, the custom InterpolationTypeError has
been dropped in favour of a TypeError with a custom message, and the
unittests have been expanded.

ambv commented on Lib/configparser.py:
```diff
@@ -612,6 +612,8 @@ def __init__(self, defaults=None, dict_type=_default_dict,
         if self._interpolation is None:
             self._interpolation = Interpolation()
+        if not isinstance(self._interpolation, Interpolation):
+            raise TypeError("interpolation is not an instance of Interpolation")
```
Since this is a debugging message, let's make it a little more verbose as to what is and isn't allowed. Plus, let's make it fit into 79 characters per PEP 8.

```suggestion
            raise TypeError(
                f"interpolation= must be None or an instance of Interpolation;"
                f" got an object of type {type(self._interpolation)}"
            )
```

ambv committed:
More verbose message
